File: 28-FindtheIndexoftheFirstOccurrenceinaString/28-FindtheIndexoftheFirstOccurrenceinaString.kt
// Last updated: 5/4/2026, 11:14:14 PM
1class Solution {
2    fun strStr(haystack: String, needle: String): Int {
3        if (needle.isEmpty()) return 0
4
5        var i = 0
6        var j = 0
7
8        while (i < haystack.length) {
9            if (haystack[i] == needle[j]) {
10                i++
11                j++
12
13                if (j == needle.length) {
14                    return i - j
15                }
16            } else {
17                i = i - j + 1
18                j = 0
19            }
20        }
21
22        return -1
23    }
24}

File: 28-FindTheIndexOfTheFirstOccurrenceInAString/28-FindTheIndexOfTheFirstOccurrenceInAString.rs
// Last updated: 8/1/2026, 11:13:29 PM
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let h = haystack.as_bytes();
        let n = needle.as_bytes();

        if n.len() == 0 {
            return 0;
        }

        if n.len() > h.len() {
            return -1;
        }

        for i in 0..=h.len() - n.len() {
            let mut match_found = true;

            for j in 0..n.len() {
                if h[i + j] != n[j] {
                    match_found = false;
                    break;
                }
            }

            if match_found {
                return i as i32;
            }
        }

        -1
    }
}